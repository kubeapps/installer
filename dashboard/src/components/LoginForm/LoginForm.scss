.LoginForm__container {
  width: 500px;
  margin: 5em auto;
}

.LoginForm > .bg-skew .bg-skew__content {
  max-width: 75%;
}
