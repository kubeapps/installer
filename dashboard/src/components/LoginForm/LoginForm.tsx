import { Location } from "history";
import * as React from "react";
import { Lock, X } from "react-feather";
import { Redirect } from "react-router";

import "./LoginForm.css";

interface ILoginFormProps {
  authenticated: boolean;
  authenticate: (token: string) => any;
  location: Location;
}

interface ILoginFormState {
  authenticating: boolean;
  token: string;
  error?: string;
}

class LoginForm extends React.Component<ILoginFormProps, ILoginFormState> {
  public state: ILoginFormState = { token: "", authenticating: false };
  public render() {
    if (this.props.authenticated) {
      const { from } = this.props.location.state || { from: { pathname: "/" } };
      return <Redirect to={from} />;
    }
    return (
      <section className="LoginForm">
        <div className="LoginForm__container padding-v-bigger bg-skew">
          <div className="container container-tiny">
            {this.state.error && (
              <div className="alert alert-error margin-c" role="alert">
                There was an error connecting to the Kubernetes API. Please check that your token is
                valid.
                <button className="alert__close" onClick={this.handleAlertClose}>
                  <X />
                </button>
              </div>
            )}
          </div>
          <div className="bg-skew__pattern bg-skew__pattern-dark type-color-reverse">
            <div className="container">
              <h2>
                <Lock /> Login
              </h2>
              <p>
                Your cluster operator should provide you with a Kubernetes API token.{" "}
                <a href="#">Click here</a> for more info on how to create and use Bearer Tokens.
              </p>
              <div className="bg-skew__content">
                <form onSubmit={this.handleSubmit}>
                  <div>
                    <label htmlFor="token">Kubernetes API Token</label>
                    <input
                      name="token"
                      id="token"
                      type="password"
                      placeholder="Token"
                      required={true}
                      onChange={this.handleTokenChange}
                      value={this.state.token}
                    />
                  </div>
                  <p>
                    <button
                      type="submit"
                      className="button button-accent"
                      disabled={this.state.authenticating}
                    >
                      Login
                    </button>
                  </p>
                </form>
              </div>
            </div>
          </div>
        </div>
      </section>
    );
  }

  private handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    this.setState({ authenticating: true });
    const { token } = this.state;
    try {
      return token && (await this.props.authenticate(token));
    } catch (e) {
      this.setState({ error: e.toString(), token: "", authenticating: false });
    }
  };

  private handleTokenChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({ token: e.currentTarget.value });
  };

  private handleAlertClose = (e: React.FormEvent<HTMLButtonElement>) => {
    this.setState({ error: undefined });
  };
}

export default LoginForm;
