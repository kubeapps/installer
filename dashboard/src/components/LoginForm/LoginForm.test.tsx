import { shallow } from "enzyme";
import { Location } from "history";
import * as React from "react";
import { Redirect } from "react-router-dom";

import LoginForm from "./LoginForm";

const emptyLocation: Location = {
  hash: "",
  pathname: "",
  search: "",
  state: "",
};

it("renders a token login form", () => {
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
  );
  expect(wrapper.find("input#token").exists()).toBe(true);
  expect(wrapper.find(Redirect).exists()).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

it("renders a link to the access control documentation", () => {
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
  );
  expect(wrapper.find("a").props()).toMatchObject({
    href: "https://github.com/kubeapps/kubeapps/blob/master/docs/user/access-control.md",
    target: "_blank",
  });
});

it("updates the token in the state when the input is changed", () => {
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
  );
  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
  expect(wrapper.state("token")).toBe("f00b4r");
});

describe("redirect if authenticated", () => {
  it("redirects to / if no current location", () => {
    const wrapper = shallow(
      <LoginForm authenticated={true} authenticate={jest.fn()} location={emptyLocation} />,
    );
    const redirect = wrapper.find(Redirect);
    expect(redirect.exists()).toBe(true);
    expect(redirect.props()).toEqual({ push: false, to: { pathname: "/" } });
  });

  it("redirects to previous location", () => {
    const location = emptyLocation;
    location.state = { from: "/test" };
    const wrapper = shallow(
      <LoginForm authenticated={true} authenticate={jest.fn()} location={location} />,
    );
    const redirect = wrapper.find(Redirect);
    expect(redirect.exists()).toBe(true);
    expect(redirect.props()).toEqual({ push: false, to: "/test" });
  });
});

it("calls the authenticate handler when the form is submitted", () => {
  const authenticate = jest.fn();
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={authenticate} location={emptyLocation} />,
  );
  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });
  expect(authenticate).toBeCalledWith("f00b4r");
  expect(wrapper.state("authenticating")).toBe(true);
});

it("displays an error if the authenticate handler throws an error", async () => {
  const authenticate = async () => {
    throw new Error("it's a trap");
  };
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={authenticate} location={emptyLocation} />,
  );
  wrapper.find("input#token").simulate("change", { currentTarget: { value: "f00b4r" } });
  wrapper.find("form").simulate("submit", { preventDefault: jest.fn() });

  // wait for promise to resolve
  try {
    await authenticate();
  } catch (e) {
    expect(wrapper.state()).toMatchObject({
      authenticating: false,
      error: e.toString(),
    });

    wrapper.update();
    expect(wrapper.find(".alert-error").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  }
});

it("allows you to dismiss the error alert", () => {
  const wrapper = shallow(
    <LoginForm authenticated={false} authenticate={jest.fn()} location={emptyLocation} />,
  );
  wrapper.setState({ error: "it's a trap" });
  expect(wrapper.find(".alert-error").exists()).toBe(true);

  wrapper.find("button.alert__close").simulate("click");
  expect(wrapper.find(".alert-error").exists()).toBe(false);
  expect(wrapper.state("error")).toBeUndefined();
});
