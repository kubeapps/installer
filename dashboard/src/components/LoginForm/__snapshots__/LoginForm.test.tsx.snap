// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`token login form displays an error if the authentication error is passed 1`] = `
<IntlProvider
  defaultFormats={Object {}}
  defaultLocale="en"
  formats={Object {}}
  locale="en"
  messages={
    Object {
      "Kubeapps": Array [
        Object {
          "type": 0,
          "value": "Kubeapps",
        },
      ],
      "Submit": Array [
        Object {
          "type": 0,
          "value": "Submit",
        },
      ],
      "Token": Array [
        Object {
          "type": 0,
          "value": "Token",
        },
      ],
      "error-login-token": Array [
        Object {
          "type": 0,
          "value": "There was an error connecting to the Kubernetes API. Please check that your token is valid.",
        },
      ],
      "login-desc-oidc": Array [
        Object {
          "type": 0,
          "value": "Your cluster operator has enabled login via an authentication provider.",
        },
      ],
      "login-desc-token": Array [
        Object {
          "type": 0,
          "value": "Your cluster operator should provide you with a Kubernetes API token.",
        },
      ],
      "login-oidc": Array [
        Object {
          "type": 0,
          "value": "Login via OIDC Provider",
        },
      ],
      "login-title-welcome": Array [
        Object {
          "type": 0,
          "value": "Welcome to",
        },
      ],
      "more-info": Array [
        Object {
          "type": 0,
          "value": "More Info",
        },
      ],
      "paste-token-here": Array [
        Object {
          "type": 0,
          "value": "Paste token here",
        },
      ],
    }
  }
  onError={[Function]}
  textComponent={Symbol(react.fragment)}
>
  <LoginForm
    appVersion="devel"
    authProxySkipLoginPage={false}
    authenticate={[MockFunction]}
    authenticated={false}
    authenticating={false}
    authenticationError="it's a trap"
    checkCookieAuthentication={[MockFunction]}
    cluster="default"
    location={
      Object {
        "hash": "",
        "pathname": "",
        "search": "",
        "state": "",
      }
    }
    oauthLoginURI=""
  >
    <div
      className="login-wrapper"
    >
      <form
        className="login clr-form"
        onSubmit={[Function]}
      >
        <TokenLogin
          authenticationError="it's a trap"
          handleTokenChange={[Function]}
          token=""
        >
          <section
            aria-describedby="login-desc"
            aria-labelledby="login-title"
            className="title"
          >
            <h3
              className="welcome"
              id="login-title"
            >
              Welcome to
              <span>
                Kubeapps
              </span>
            </h3>
            <p
              className="hint"
              id="login-desc"
            >
              Your cluster operator should provide you with a Kubernetes API token.
            </p>
            <div
              className="login-group"
            >
              <div
                className="clr-form-control"
              >
                <label
                  className="clr-control-label"
                  htmlFor="token"
                >
                  Token
                </label>
                <div
                  className="clr-control-container"
                >
                  <div
                    className="clr-input-wrapper"
                  >
                    <input
                      className="clr-input"
                      id="token"
                      name="token"
                      onChange={[Function]}
                      placeholder="Paste token here"
                      required={true}
                      type="password"
                      value=""
                    />
                  </div>
                </div>
              </div>
              <div
                className="error active "
              >
                There was an error connecting to the Kubernetes API. Please check that your token is valid.
              </div>
              <div
                className="login-submit-button"
              >
                <CdsButton
                  id="login-submit-button"
                  status="primary"
                >
                  <cds-button>
                    Submit
                  </cds-button>
                </CdsButton>
              </div>
            </div>
          </section>
        </TokenLogin>
        <div
          className="login-moreinfo"
        >
          <a
            href="https://github.com/kubeapps/kubeapps/blob/devel/docs/user/access-control.md"
            rel="noopener noreferrer"
            target="_blank"
          >
            <CdsIcon
              shape="info-circle"
            >
              <cds-icon />
            </CdsIcon>
            More Info
          </a>
        </div>
      </form>
    </div>
  </LoginForm>
</IntlProvider>
`;

exports[`while authenticating loading spinner matches the snapshot 1`] = `
<LoadingWrapper
  loaded={false}
/>
`;
